Better errors when app command fails

This is what it looks like currently:

```
$ yarn test
yarn run v1.22.4
$ bigtest server
[orchestrator] starting
Error: 'yarn start' exited with non-zero exit code
    ... useless stack trace ...
error Command failed with exit code 255.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

We should surface the underlying error. In my case the error was this, which was an easy fix:

```
$ yarn start
yarn run v1.22.4
$ ts-node src/index.ts
Error: Cannot find module '/Users/rpitts/.../index.ts'
    ... useless stack trace ...
```

--- a/.changeset/cyan-pandas-think.md
+++ b/.changeset/cyan-pandas-think.md
@@ -1,0 +1,6 @@
+---
+"@bigtest/server": minor
+"bigtest": minor
+---
+
+Show app errors on test run if application server exits prematurely

--- a/packages/cli/test/cli.test.ts
+++ b/packages/cli/test/cli.test.ts
@@ -268,6 +268,23 @@ describe('@bigtest/cli', function() {
         expect(child.stdout?.output).toContain('⨯ FAILURE')
       });
     });
+
+    describe('running the suite with app errors', () => {
+      let child: TestProcess;
+      let status: ExitStatus;
+
+      beforeEach(async () => {
+        child = await run('ci', '--app-command', 'yarn node doesnotexist.js');
+        status = await child.join();
+      });
+
+      it('exits with error code', async () => {
+        expect(status.code).toEqual(1);
+        expect(child.stdout?.output).toContain('Application exited unexpectedly with exit code 1 with the following output:')
+        expect(child.stdout?.output).toContain('Cannot find module')
+        expect(child.stdout?.output).toContain('⨯ FAILURE')
+      });
+    });
   });
 
   describe('coverage', () => {

--- a/packages/server/src/app-server.ts
+++ b/packages/server/src/app-server.ts
@@ -33,21 +33,19 @@ const startApp = ({ atom }: AppServerOptions): Service<AppOptions> => function* 
     yield spawn(function* () {
       let exitStatus = yield child.join();
 
-      appStatus.set({ type: 'crashed', exitStatus });
+      appStatus.set({ type: 'exited', exitStatus });
     });
   }
 
   appStatus.set({ type: 'started' });
-  
-  while(true) {
+
+  while(!(yield isReachable(options.url))) {
     yield timeout(100);
-    
-    if (yield isReachable(options.url)) {
-      appStatus.set({ type: 'reachable' });
-    } else {
-      appStatus.set({ type: 'unreachable' });
-    }
   }
+
+  appStatus.set({ type: 'reachable' });
+
+  yield;
 }
 
 function* isReachable(url: string) {

--- a/packages/server/src/command-processor.ts
+++ b/packages/server/src/command-processor.ts
@@ -47,6 +47,21 @@ function* run({ id: testRunId, files }: RunMessage, options: CommandProcessorOpt
       return;
     }
 
+    let appStatus = options.atom.slice("appService", "status").get();
+
+    if(appStatus.type === 'exited') {
+      testRunSlice.set({
+        testRunId: testRunId,
+        status: 'failed',
+        error: {
+          name: 'AppError',
+          message: `Application exited unexpectedly with exit code ${appStatus.exitStatus.code} with the following output:\n\n${appStatus.exitStatus.tail}`
+        },
+        agents: {}
+      });
+      return;
+    }
+
     // todo: we should perform filtering of the agents here
     let agents: AgentState[] = Object.values(options.atom.get().agents);
 

--- a/packages/server/src/logger.ts
+++ b/packages/server/src/logger.ts
@@ -1,3 +1,4 @@
+import { fork } from 'effection';
 import { OrchestratorState } from './orchestrator/state';
 import { Atom } from '@bigtest/atom';
 import { subscribe } from '@effection/subscription';
@@ -8,9 +9,7 @@ export interface LoggerOptions {
 }
 
 export function* createLogger({ atom, out }: LoggerOptions) {
-  let bundlerState = atom.slice('bundler');
-
-  yield subscribe(bundlerState).forEach(function* (event) {
+  yield fork(subscribe(atom.slice('bundler')).forEach(function* (event) {
     if(event.type === 'ERRORED'){
       out("[manifest builder] build error:");
       out(event.error.message);
@@ -18,5 +17,14 @@ export function* createLogger({ atom, out }: LoggerOptions) {
     if(event.type === 'GREEN'){
       out("[manifest builder] build successful!");
     }
-  })
+  }));
+
+  yield fork(subscribe(atom.slice('appService', 'status')).forEach(function* (status) {
+    if(status.type === 'reachable') {
+      out("[app] successfully connected to application!");
+    }
+    if(status.type === 'exited') {
+      out(`[app] application has exited with status code ${status.exitStatus.code}`)
+    }
+  }));
 }

--- a/packages/server/src/orchestrator.ts
+++ b/packages/server/src/orchestrator.ts
@@ -119,10 +119,10 @@ export function* createOrchestrator(options: OrchestratorOptions): Operation {
       console.debug('[orchestrator] connection server ready');
     });
     yield fork(function*() {
-      yield options.atom.slice('appService', 'status').once((status) => {
-        return status.type === 'reachable'
+      let status = yield options.atom.slice('appService', 'status').once((status) => {
+        return status.type === 'reachable' || status.type === 'exited';
       });
-      console.debug('[orchestrator] app server ready');
+      console.debug(`[orchestrator] app server ${status.type}`);
     });
     yield fork(function* () {
       yield options.atom.slice('bundler').once(({ type }) => type === 'GREEN' || type === 'ERRORED');

--- a/packages/server/src/orchestrator/state.ts
+++ b/packages/server/src/orchestrator/state.ts
@@ -51,11 +51,10 @@ export type BundlerState =
 export type BundlerTypes = Pick<BundlerState, 'type'>['type'];
 
 export type ServiceStatus =
-  | { type: 'unstarted' } 
+  | { type: 'unstarted' }
   | { type: 'started' }
   | { type: 'reachable' }
-  | { type: 'unreachable' }
-  | { type: 'crashed'; exitStatus: ExitStatus };
+  | { type: 'exited'; exitStatus: ExitStatus };
 
 export type ServiceStatuses = ServiceStatus['type'];
 

--- a/packages/server/test/app-server.test.ts
+++ b/packages/server/test/app-server.test.ts
@@ -16,56 +16,50 @@ describe('app service', () => {
       atom = createOrchestratorAtom({ app: {
           url: "http://localhost:24000",
           command: "yarn test:app:start 24000",
-        } 
+        }
       });
-  
+
       actions.fork(function * () {
         yield createAppServer({ atom })
       });
     });
-  
-    it("should be transition from 'started' to 'unreachable' to 'reachable'", async () => {
+
+    it("should be transition from 'started' to 'reachable'", async () => {
       let appStatus = atom.slice('appService', 'status');
-  
+
       expect(appStatus.get().type).toBe('started');
-  
-      await actions.fork(
-        appStatus.once(status => status.type === 'unreachable')
-      );
-  
-      expect(appStatus.get().type).toBe('unreachable');
-  
+
       await actions.fork(
         appStatus.once(status => status.type === 'reachable')
       );
-  
+
       expect(appStatus.get().type).toBe('reachable');
     });
   });
 
-  describe('crashed', () => {
+  describe('exited', () => {
     beforeEach(() => {
       atom = createOrchestratorAtom({ app: {
           url: "http://localhost:24000",
           command: "yarn no:such:command",
-        } 
+        }
       });
-  
+
       actions.fork(function * () {
         yield createAppServer({ atom })
       });
     });
 
-    it('should transition to crashed', async () => {
+    it('should transition to exited', async () => {
       let appStatus = atom.slice('appService', 'status');
 
       await actions.fork(
-        appStatus.once(status => status.type === 'crashed')
+        appStatus.once(status => status.type === 'exited')
       );
 
       let current = appStatus.get();
 
-      assertAppServiceStatus(current.type, { is: 'crashed' });
+      assertAppServiceStatus(current.type, { is: 'exited' });
 
       expect(current.exitStatus.code).toBe(1);
     })

